### PR TITLE
Add api to control thread for all index

### PR DESCRIPTION
**Pull request by smellthemoon**
Signed-off-by: lixinguo <xinguo.li@zilliz.com>
issue: #374 

It offers api to control threads when query and build in all indexes. You can add build_thread_num and query_thread_num when creating config and it offers the default values if you don't add it.

### CODE REVIEW

**Review by cydrain — COMMENTED** [review]
if (CheckKeyInConfig(config, meta::QUERY_THREAD_NUM)) {
            omp_set_num_threads(GetMetaQueryThreadNum(config));
        } else {
            omp_set_num_threads(knowhere::DEFAULT_QUERY_THREAD_NUM);
        }

File: knowhere/index/vector_index/helpers/IndexParameter.h [hunk 1]
```diff
@@ -39,6 +39,9 @@ constexpr const char* BUILD_THREAD_NUM = "build_thread_num";
 constexpr const char* QUERY_THREAD_NUM = "query_thread_num";
```
Comment (cydrain) [inline knowhere/index/vector_index/helpers/IndexParameter.h:39]: rename to "BUILD_INDEX_OMP_NUM" and "QUERY_OMP_NUM"

**Comment by liliu-z** [comment]
> Following code piece is repeated more than 10 times.
> 
> ```
>         if (CheckKeyInConfig(config, meta::QUERY_THREAD_NUM)) {
>             omp_set_num_threads(GetMetaQueryThreadNum(config));
>         } else {
>             omp_set_num_threads(knowhere::DEFAULT_QUERY_THREAD_NUM);
>         }
> ```
> 
> Suggest: add new files "omp.h" and "omp.cpp" under knowhere/common, and create an new API
> 
> * SetOmpWithDefault(const Config& conf, const MetaType& type, const int64_t default_val)

Good catch, I thought `Utils.h` under `common` folder sounds more rational to me. We will not have too many `omp` related common functions

File: knowhere/index/vector_offset_index/IndexIVF_NM.cpp [hunk 2]
```diff
@@ -119,8 +119,11 @@ IVF_NM::Train(const DatasetPtr& dataset_ptr, const Config& config) {
     auto coarse_quantizer = new faiss::IndexFlat(dim, metric_type);
     auto index = std::make_shared<faiss::IndexIVFFlat>(coarse_quantizer, dim, nlist, metric_type);
     index->own_fields = true;
-    if (CheckKeyInConfig(config, meta::BUILD_THREAD_NUM))
+    if (CheckKeyInConfig(config, meta::BUILD_THREAD_NUM)) {
         omp_set_num_threads(GetMetaBuildThreadNum(config));
+    } else {
+        omp_set_num_threads(knowhere::DEFAULT_BUILD_THREAD_NUM);
```
Comment (cydrain) [inline knowhere/index/vector_offset_index/IndexIVF_NM.cpp:125]: default omp num should be std::min(system_max_omp, DEFAULT_XXX_OMP_NUM)

File: knowhere/index/vector_index/IndexAnnoy.cpp [hunk 3]
```diff
@@ -141,7 +143,11 @@ IndexAnnoy::Query(const DatasetPtr& dataset_ptr, const Config& config, const fai
     auto search_k = GetIndexParamSearchK(config);
     auto p_id = new int64_t[k * rows];
     auto p_dist = new float[k * rows];
-
+    if (CheckKeyInConfig(config, meta::QUERY_THREAD_NUM)) {
+        omp_set_num_threads(GetMetaQueryThreadNum(config));
```
Comment (smellthemoon) [inline knowhere/index/vector_index/IndexAnnoy.cpp:147]: It set two parameters in meta,QUERY_THREAD_NUM and BUILD_THREAD_NUM, which refers to the parameters that user assigned. When users didn't assign it, the omp parallel section will use the default parameter seted in knowhere, DEFUALT_QUERY_THREAD_NUM and DEFUALT_BUILD_THREAD_NUM.

**Pull request closed without merging**